### PR TITLE
[SWY-118] Song resource empty state UI

### PR DESCRIPTION
**Pull request by justaddcl**
## Summary

[SWY-118]

- Add an illustration for the empty state for songs with no linked resources.
- Wire the empty-state "Link a resource" CTA to the existing create resource dialog.
- Move resource forms to URL-first field order for create and edit flows.
- Improve resource form validation timing:
  - required-field errors appear on submit, not on initial blur
  - non-HTTPS URL errors appear after the URL field has a value and blurs
  - clearing the URL field clears URL format errors until submit
  - whitespace-only URLs are treated as missing values
- Fix edit-mode no-op saves by keeping "Save changes" disabled until the form is dirty.
- Improve mobile drawer/input behavior:
  - prevent iOS input zoom with 16px mobile input text
  - align the drawer with Vaul's input handling guidance
  - disable Vaul input repositioning by default
  - lock background page scrolling while the drawer is open
  - keep long drawer content scrollable inside the sheet
- Optimize the empty-state illustration with a WebP asset and intrinsic image sizing.
- Clean up the song resources query state now that the empty state is implemented.

| Mobile | Desktop |
| ------ | ------- |
| SWY-118__after--mobile | SWY__after--desktop |

## Why

Songs without linked resources now have a polished, production-ready empty state that gives users a clear next action. The related form and mobile drawer fixes make that action usable across desktop and mobile, especially on iOS where input focus can otherwise zoom or move the drawer/page unexpectedly.

## Notable implementation details

- `SongResourcesEmptyState` renders the new empty-state list item and opens the existing `ResponsiveDialog` through the parent `SongResources` state.
- `ResourceForm` now uses `mode: "onSubmit"` with custom URL `onChange`/`onBlur` handling so the form is less aggressive about empty required fields while still catching invalid URL schemes early.
- `resourceUrlSchema` trims URL input before required-field validation and keeps the HTTPS/domain validation in the shared Zod schema.
- `Drawer` defaults to `repositionInputs={false}` and `noBodyStyles={true}`, then applies a local ref-counted scroll lock so the page behind the drawer does not scroll while the drawer is open.
- `Input` uses `text-base md:text-sm`, keeping mobile input text at 16px to avoid iOS Safari auto-zoom while preserving the desktop text size.
- `SongResources` now removes the unused resource error destructure/comment left over from the previous placeholder state.
- `SongResourcesEmptyStateBackupArt` remains exported as an unused packet-art fallback.

## Validation

- `pnpm test:jest src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx --runInBand`
- `pnpm exec eslint src/components/ui/drawer.tsx src/components/ui/input.tsx src/lib/types/zod.ts src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx src/modules/songs/components/SongResources/SongResources.tsx src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx`
  - passes with the existing `ZodIssueCode` deprecation warning in `src/lib/types/zod.ts`
- `pnpm exec tsc --noEmit --project tsconfig.json`
- `git diff --check`

CI status is being refreshed for the latest cleanup commit.

<h3>Greptile Summary</h3>

This PR delivers the production empty state for songs with no linked resources, wires the "Link a resource" CTA to the existing `ResponsiveDialog`, and ships a set of mobile/form polish fixes — iOS input zoom prevention, Vaul drawer scroll-lock, URL-first field order, and refined validation timing.

- **Empty state**: New `SongResourcesEmptyState` component renders the WebP glow illustration and a `Button` that opens the link-resource dialog via `handleAddResourceClick` hoisted in `SongResources`.
- **Form validation**: `ResourceForm` moves to `mode:"onSubmit"` / `reValidateMode:"onChange"` with custom URL `onChange`/`onBlur` handlers; `resourceUrlSchema` gains `.trim().min(1)` so whitespace-only input shows a "required" error rather than a format error; edit-mode submit is gated on `isDirty`.
- **Drawer / Input**: Custom ref-counted `lockPageScroll`/`unlockPageScroll` module-level helpers prevent background scroll when Vaul's body styles are disabled; `Input` switches to `text-base md:text-sm` to suppress iOS auto-zoom.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all changed paths are UI-layer only with no data-mutation logic changes, and the new validation schema is strictly additive.

The scroll-lock implementation is well-guarded (ref-counted, only runs inside useLayoutEffect on the client), the Zod schema change correctly adds trim+min before the URL pipe, and the form disabled-logic correctly distinguishes create vs edit mode. The only non-ideal point is SongResourcesEmptyState lacking an explicit "use client" boundary, which is harmless today but worth a follow-up.

src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx — missing "use client" directive.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx | New component: renders the glow-art illustration and "Link a resource" CTA inside a list-item shell. Missing "use client" directive despite containing an onClick handler; safe for now due to parent boundary but fragile. |
| src/components/ui/drawer.tsx | Adds ref-counted module-level scroll lock (lockPageScroll/unlockPageScroll), wires it to a usePageScrollLock hook inside the Drawer wrapper, and sets noBodyStyles=true and repositionInputs=false as defaults. DrawerContent switched to max-h-[82vh] with overflow-hidden and uses DrawerPrimitive.Handle instead of a manual pill div. |
| src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx | Switches to mode:"onSubmit"/reValidateMode:"onChange", swaps field order to URL-first, adds custom onChange/onBlur handlers to give URL-specific validation timing, and fixes edit-mode disabled logic to gate on isDirty. |
| src/lib/types/zod.ts | Replaces bare z.url() in resourceUrlSchema with z.string().trim().min(1).pipe(z.url(…)), adding an explicit required-field message and whitespace trimming before URL format validation. |
| src/modules/songs/components/SongResources/SongResources.tsx | Replaces inline "No song resources yet" placeholder with SongResourcesEmptyState, wires its CTA to the existing ResponsiveDialog via handleAddResourceClick, and removes the consumed songResourcesError binding (leaving a TODO for error state). |
| src/modules/songs/components/SongResources/__tests__/SongResources.test.tsx | Adds coverage for the empty-state CTA, URL-first field order, required-field validation timing, whitespace URL handling, https-only validation clearing, and edit-mode dirty gating. |
| src/components/ui/input.tsx | Swaps text-sm for text-base md:text-sm to keep mobile input text at 16px and avoid iOS Safari auto-zoom. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant SongResourcesEmptyState
    participant SongResources
    participant ResponsiveDialog
    participant ResourceForm
    participant Zod as resourceUrlSchema

    User->>SongResourcesEmptyState: click "Link a resource"
    SongResourcesEmptyState->>SongResources: onAddResourceClick()
    SongResources->>SongResources: setIsLinkResourceDialogOpen(true)
    SongResources->>ResponsiveDialog: "open=true"
    ResponsiveDialog->>ResourceForm: "render (mode=create)"

    User->>ResourceForm: fill URL field
    ResourceForm->>ResourceForm: "onChange → trigger url if errors or submitCount>0"
    User->>ResourceForm: blur URL field
    ResourceForm->>ResourceForm: onBlur → trigger url if value non-empty

    User->>ResourceForm: submit
    ResourceForm->>Zod: validate url and title
    alt URL empty or whitespace
        Zod-->>ResourceForm: Please enter a resource URL
        ResourceForm-->>User: show error, button disabled
    else URL non-https
        Zod-->>ResourceForm: Please use a link starting with https
        ResourceForm-->>User: show error, button disabled
    else valid
        ResourceForm->>ResourceForm: createResourceMutation.mutateAsync()
        ResourceForm-->>SongResources: onSuccess()
        SongResources->>ResponsiveDialog: close
    end
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx:1
**Missing `"use client"` directive**

`SongResourcesEmptyState` renders an interactive `Button` with an `onClick` handler, but the file has no `"use client"` boundary. In Next.js App Router, a file without this directive is treated as a server component by default. It works today only because it is transitively imported by `SongResources.tsx` (which carries `"use client"`), but if this component is ever imported directly in a server-component context — or moved to a shared location — the event handler will fail at runtime.

`````

</details>

<sub>Reviews (9): Last reviewed commit: ["Clean up song resources query state"](https://github.com/justaddcl/sanbi/commit/b04be40ee9edc7e21baf54ae2e66809371482504) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=33095354)</sub>

### CODE REVIEW

**Comment by justaddcl** [comment]
#### 📝 Changed routes:


| [**/**](https://sanbi-1u1s8lzlx-justaddcls-projects.vercel.app/) |
|:-:|
|<a href="https://sanbi-1u1s8lzlx-justaddcls-projects.vercel.app/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/IytJAmvkfAAjDce7TWW0HnHyS80htcRT73K6JhdNyZqzTsxjtDn9zwyQqENATzJEB-B1tJtZDhC8ldJMYcZXuQ.png">(view full size)</a> |







Commit 13e18f9bfbfeacfbfc4c838aa03b2e443e45729f (https://sanbi-1u1s8lzlx-justaddcls-projects.vercel.app).

File: src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  </div>
+);
+
+const LinkedGlowArt = () => (
+  <img
+    src="/song-resource-empty-states/linked-glow.png"
+    alt=""
+    aria-hidden
+    className="h-40 w-full max-w-md shrink-0 object-contain object-center"
+  />
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/components/SongResources/SongResourcesEmptyState.tsx]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by adding a WebP version of the empty-state art, switching the component to `next/image`, and providing intrinsic dimensions plus lazy loading.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 2]
```diff
@@ -227,23 +228,50 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
...
+                    }}
+                    onBlur={(event) => {
+                      field.onBlur();
+
+                      if (event.target.value.trim() || submitCount > 0) {
+                        void trigger("url");
+                        return;
+                      }
+
+                      clearErrors("url");
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx:263]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by dropping the `submitCount > 0` branch from both URL change/blur validation paths, so clearing the URL clears URL errors until the next submit.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx [hunk 3]
```diff
@@ -213,7 +213,8 @@ export const ResourceForm: React.FC<ResourceFormProps> = ({
     }
   };
 
-  const shouldSubmitButtonBeDisabled = !isDirty || !isValid || isSubmitting;
+  const shouldSubmitButtonBeDisabled =
+    isSubmitting || Object.keys(errors).length > 0;
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by gating edit-mode submission on `isDirty` while leaving create-mode submit available for empty-form validation.

File: src/components/ui/drawer.tsx [hunk 4]
```diff
@@ -53,13 +50,13 @@ const DrawerContent = React.forwardRef<
...
+        "fixed inset-x-0 bottom-0 z-50 flex max-h-[82vh] flex-col rounded-t-[10px] border bg-background",
         className,
       )}
       {...props}
     >
-      <div className="flex-1 overflow-y-auto p-6">
-        <div className="mx-auto mt-4 h-2 w-[100px] rounded-full bg-muted" />
+      <div className="mx-auto w-full max-w-md overflow-auto rounded-t-[10px] p-4 pb-[calc(1rem+env(safe-area-inset-bottom))]">
+        <DrawerPrimitive.Handle />
         {children}
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/components/ui/drawer.tsx]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by hiding overflow on the drawer content shell and making the inner content container a constrained `min-h-0 flex-1 overflow-y-auto` scroller.

File: src/lib/types/zod.ts [hunk 5]
```diff
@@ -297,11 +297,16 @@ export const getResourcesBySongIdSchema = z.object({
...
 
-const resourceUrlSchema = z.url({
-  protocol: /^https$/,
-  hostname: z.regexes.domain,
-  error: "Please use a link starting with https://",
-});
+const resourceUrlSchema = z
+  .string()
+  .min(1, "Please enter a resource URL")
+  .pipe(
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/types/zod.ts:304]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by trimming `resourceUrlSchema` before the required-field check. I also added a workflow test covering whitespace-only URL submission.

File: src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/CreateResourceForm/CreateResourceForm.tsx]: Addressed in https://github.com/justaddcl/sanbi/commit/9a37b65 by restoring edit-mode dirty-state gating and adding a test that `Save changes` stays disabled until a field changes.

File: src/components/ui/input.tsx [hunk 6]
```diff
@@ -15,7 +15,7 @@ export const Input = React.forwardRef<HTMLInputElement, InputProps>(
       <input
         type={type}
         className={cn(
-          "flex h-10 w-full rounded-md border border-input bg-background py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50",
+          "flex h-10 w-full rounded-md border border-input bg-background py-2 text-base ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/components/ui/input.tsx:18]: I’m going to keep this scoped in the shared `Input` primitive. This repository’s shadcn-style primitives are customized directly in component source, and the 16px mobile font size is part of this primitive’s behavior to avoid iOS input zoom. Moving it to a global `input` selector would be broader and easier to apply to native inputs that are not this component.

File: src/app/layout.tsx [hunk 7]
```diff
@@ -38,7 +38,10 @@ export default function RootLayout({
     <ClerkProvider>
       <html lang="en" className={`${poppins.variable}`}>
         <head>
-          <meta name="viewport" content="width=device-width, initial-scale=1" />
+          <meta
+            name="viewport"
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/app/layout.tsx]: This is already addressed by https://github.com/justaddcl/sanbi/commit/4d1203d. The current viewport meta is `width=device-width, initial-scale=1`, so intentional user zoom is no longer suppressed.
